File: 0803-cheapest-flights-within-k-stops/0803-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int K) {
        vector<int>next(n, INT_MAX/2);
        next[src] = 0;

        for (int k = 0; k < K+1; k++){
            vector<int> dp_temp(next);
            for (auto ticket: flights){
                next[ticket[1]] = min(next[ticket[1]], dp_temp[ticket[0]]+ticket[2]);
            }
        }
        if (next[dst] != INT_MAX/2){
            return next[dst];
        }
        return -1;

    }
};

File: 0803-cheapest-flights-within-k-stops/0803-cheapest-flights-within-k-stops.py
import heapq
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        next = defaultdict(list)

        for x,y,z in flights:
            next[x].append((y,z))

        cost = [[float('inf') for _ in range(k+2)] for _ in range(n)]

        pq = [(0 , src, 0)]

        while pq:
            c, cur, stops = heapq.heappop(pq)

            if cur == dst:
                return c 
            
            if stops == k + 1:
                continue

            if cost[cur][stops] < float('inf'):
                continue
            
            cost[cur][stops] = c

            for nxt, price in next[cur]:
                if cost[nxt][stops+1] < float('inf'):
                    continue
                heapq.heappush(pq, (c+price, nxt, stops+1))
        
        return -1 